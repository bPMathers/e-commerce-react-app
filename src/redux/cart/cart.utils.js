// .utils file is used to separate logic from the reducer file to make it clearer to read and test


export const addItemToCart = (cartItems, cartItemToAdd) => {
  const existingCartItem = cartItems.find(cartItem => cartItem.id === cartItemToAdd.id);

  if (existingCartItem) {
    return cartItems.map(cartItem =>
      cartItem.id === cartItemToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    )
  }

  return [...cartItems, { ...cartItemToAdd, quantity: 1 }]
};