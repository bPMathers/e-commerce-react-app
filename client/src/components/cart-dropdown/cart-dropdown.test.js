import React from 'react';
import { shallow } from 'enzyme';

import { CartDropdown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';
import { toggleCartHidden } from '../../redux/cart/cart.actions';


describe('CartDropdown component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn()
    };
    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it('should render CartDropdown component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should navigate to checkout and dipatch toggleCartHidden() when clicking CartDropdownButton ', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalledWith('/checkout');
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('should render render an equal number of cartItem components as the cartItems prop contain', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length)
  });

  it('should render an empty message container when cart is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch
    };
    const wrapper2 = shallow(<CartDropdown {...mockProps} />);
    expect(wrapper2.exists('EmptyMessageContainer')).toBe(true);
  })
});