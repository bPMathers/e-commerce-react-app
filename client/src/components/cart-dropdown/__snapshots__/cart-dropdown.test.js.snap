// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should render CartDropdown component 1`] = `
<CartDropdownContainer>
  <CartItemsContainer>
    <CartItem
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <CartItem
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <CartItem
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </CartItemsContainer>
  <CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropdownButton>
</CartDropdownContainer>
`;
